shell/common/engine.cc

```diff
@@ -103,6 +103,7 @@ bool Engine::Restart(RunConfiguration configuration) {
   }
   runtime_controller_ = runtime_controller_->Clone();
   UpdateAssetManager(nullptr);
+  delegate_.OnEngineRestart();
```

Comment:
You need to check if the run call actually succeeded before firing the delegate method. As written failed attempts to relaunch the engine will still fire the delegate callback.

Reply:
That might be a problem - I need to make sure I flushed the embedded views before the dart code starts...
What happens when a hot restart fails? do we retry? what if all retries failed?

Reply:
When a hot restart fails, we just sit there waiting for another restart (all restarts are initiated by the service protocol).

I you really want a callback just before the restart, I would name this delegate method `OnEngineWillRestart` to better indicate that the restart has not yet happened. Also, move this delegate call up before the `runtime_controller_->Clone()` call. On line 106, the old runtime controller has already been destroyed.